crates/core/component/dex/src/component/router/fill_route.rs

```diff
@@ -380,17 +379,28 @@ impl<S: StateRead + StateWrite> Frontier<S> {
     /// Apply the [`FrontierTx`] to the frontier, returning the input and output
     /// amounts it added to the trace.
     fn apply(&mut self, changes: FrontierTx) -> (Amount, Amount) {
-        self.trace[0].amount +=
-            changes.trace[0].expect("all trace amounts must be set when applying changes");
+        let mut trace: Vec<Value> = vec![];
+
+        trace[0] = Value {
+            amount: changes.trace[0].expect("all trace amounts must be set when applying changes"),
+            asset_id: self.trace[0][0].asset_id,
```

Comment:
Rather than pulling the asset ID out of `self.trace[0]`, which may not be present, they can be pulled out of `self.pairs` (`.start` for `trace[0]` and `.end` for all other `trace[i]`).